docs/developer/explanations/type_validators.rst

```diff
@@ -52,55 +52,48 @@ to be devices and look them up against the context. For example with the request
 
 Solution
 --------
-
-Before pydantic, blueapi used apischema_, which had an ideal feature for this called conversions_.
-Currently in the utils module of a blueapi is a similar feature called type validators.
-
-They enable the ``BlueskyContext`` to dynamically generate pydantic models, like above, that look 
-roughly like this:
+When the context loads available plans, it iterates through the type signature
+and replaces any reference to a bluesky protocol (or instance of a protocol)
+with a new class that extends the original type. Defining a class validator on
+this new type allows it to check that the string being deserialised is the ID of
+a device of the correct type.
+
+These new intermediate types are used only in the deserialisation process. The
+object returned from validator method is not checked by pydantic so it can be
+the actual instance and the plan never sees the runtime generated reference
+type, only the type it was expecting.
+
+NB. This uses the fact that the new types generated at runtime have access to
+the context that required them via their closure. This circumvents the usual
+problem of pydantic validation not being able to access external state when
+validating or deserialising.
 
 .. code:: python
-    
+
     def my_plan(a: int, b: Readable) -> Plan:
         ...
 
     # Becomes
 
     class MyPlanModel(BaseModel):
         a: int
-        b: Readable
+        b: Reference[Readable]
 
-        @validator("b")
-        def valdiate_b(self, val: str) -> Readable:
-            return ctx.find_device(val)
 
-
-It also handles the case of the ``Readable`` type being placed at various type levels? For example:
+This also allows ``Readable`` to be placed at various type levels. For example:
 
 .. code:: python
-    
+
     def my_weird_plan(
-        a: Readable, 
-        b: List[Readable], 
-        c: Dict[str, Readable], 
-        d: List[List[Readable]], 
+        a: Readable,
+        b: List[Readable],
+        c: Dict[str, Readable],
+        d: List[List[Readable]],
         e: List[Dict[str, Set[Readable]]]) -> Plan:
         ...
 
 
 Implementation Details
```

Comment:
I think the referencing is explained sufficiently above, so can lose the heading.